addons/website_theme_install/models/ir_module_module.py

```diff
@@ -355,10 +349,11 @@ def button_choose_theme(self):
         # website.theme_id must be set before upgrade/install to trigger the load in ``write``
         website.theme_id = self
 
-        self._theme_upgrade_upstream()
+        def install_or_upgrade(theme):
+            theme._theme_prepare_upgrade_upstream()
+            theme.button_install()
 
-        if self.state != 'installed':
-            self.button_immediate_install()
+        self._button_immediate_function(install_or_upgrade)
```

Comment:
To check:

Couldn't we just call `_theme_upgrade_upstream` here, and add `theme.button_install()` in `_theme_prepare_upgrade_upstream` (and renaming it) ?

If the condition on `self.state != 'installed'` isn't necessary, the code could be the same in both cases.

Reply:
@seb-odoo you are right, but this actually changes the semantics of `_theme_upgrade_upstream`, and we should *rename* the method to avoid any confusion...

Reply:
@seb-odoo I just added a commit to test your idea.
It's actually not bad, but the name of the method is a bit confusing.

Reply:
Well I don't think we can change the name of the method in stable, but your comment seems good enough to me.